src/peer_select.cc

```diff
@@ -99,6 +101,156 @@ operator <<(std::ostream &os, const PeerSelectionDumper &fsd)
+/// Each map element is a (pingTime, PeerSelector) pair, where pingTime is an
+/// expiration time, i.e., ping reply absolute time maximum for the given PeerSelector.
+/// Scheduling only one event at a time (for PeerSelector with the lowest expiration time),
+/// it helps to optimize situations when there are many (thousands) of concurrent
+/// transactions: an inefficient alternative would add thousands of concurrent events
+/// to the Squid event queue.
+class PeerSelectorTimeoutProcessor
+{
+public:
+    static void NoteWaitOver(void *raw);
+
+    /// calls back all ready PeerSelectors and continues to wait for others
+    void noteWaitOver();
+
+    /// adds a PeerSelector to the map
+    void enqueue(PeerSelector *selector);
+
+    /// removes a PeerSelector from the map
+    void dequeue(PeerSelector *selector);
+
+    /// \returns the 'end' iterator of the map
+    PeerSelectorMapIterator end() { return waitingMap.end(); }
+
+    /// the scheduled event absolute time
+    timeval nextEventTime() const;
+
+private:
+    void addEvent();
+    void deleteEvent();
+
+    PeerSelectorMap waitingMap; ///< postponed PeerSelectors as (timeval, PeerSelector*) pairs
+};
+
+PeerSelectorTimeoutProcessor ThePeerSelectorTimeoutProcessor;
+
+timeval
+PeerSelectorTimeoutProcessor::nextEventTime() const
+{
+    Must(!waitingMap.empty());
+    return waitingMap.begin()->first;
+}
+
+void
+PeerSelectorTimeoutProcessor::NoteWaitOver(void *raw)
+{
+    assert(raw);
+    static_cast<PeerSelectorTimeoutProcessor*>(raw)->noteWaitOver();
+}
+
+static double
+DayTimeOf(const PeerSelectorMapIterator &it)
+{
+    Must(it != ThePeerSelectorTimeoutProcessor.end());
+    // convert timeval into the time compatible with current_dtime
+    return static_cast<double>(it->first.tv_sec) + static_cast<double>(it->first.tv_usec) / 1000000.0;
+}
+
+void
+PeerSelectorTimeoutProcessor::addEvent()
+{
+    Must(!waitingMap.empty());
+    const auto interval = DayTimeOf(waitingMap.begin()) - current_dtime;
+    eventAdd("PeerSelectorTimeoutProcessor::NoteWaitOver", &PeerSelectorTimeoutProcessor::NoteWaitOver, this, interval, 0, false);
+}
+
+void
+PeerSelectorTimeoutProcessor::deleteEvent()
+{
+    eventDelete(&PeerSelectorTimeoutProcessor::NoteWaitOver, nullptr);
+}
+
+static void
+HandlePingTimeout(PeerSelector *selector)
+{
+    selector->handlePingTimeout();
+}
+
+void
+PeerSelectorTimeoutProcessor::noteWaitOver()
+{
+    if (waitingMap.empty())
+        return;
+
+    (void)getCurrentTime();
+
+    auto it = waitingMap.begin();
+    for (; it != waitingMap.end(); ++it)
+    {
+        if (DayTimeOf(it) > current_dtime)
+            break;
+        const auto selector = it->second;
+        CallBack(selector->al, [selector] {
+            AsyncCall::Pointer callback = asyncCall(44, 4, "HandlePingTimeout", cbdataDialer(HandlePingTimeout, selector));
+            ScheduleCallHere(callback);
+        });
+    }
+
+    waitingMap.erase(waitingMap.begin(), it);
```

Comment:
The combination of this noteWaitOver() code and stopPingWaiting() that this code (indirectly) asynchronously calls creates a dangerous state where entry's ping.waitPosition is essentially invalid but the corresponding peer selector is still in "waiting" state. We should fix this IMO. I suggest simplifying this loop by always checking (and, if necessary, removing) the first element. It may be a bit more expensive than erasing a range, but it is clearer and safer. Here is a sketch:

```C++
while (!waitingMap.empty() && front().ping.timedOut()) {
    schedule an async call
    reset front().ping's iterator
    waitingMap.erase(waitingMap.begin());
}
```